Bottom: 157e96e989dddeb2c576dffb74a1dfd65d0d4458
Top:    e6b2f720094da0180d95d8f32a8a3abc93bba8c9
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-08-16 09:17:44 -0500

Update ptxdist build factories


---
diff --git a/roles/buildbot-master/files/ptxdist.py b/roles/buildbot-master/files/ptxdist.py
index d6d84f2e..13e9b27e 100644
--- a/roles/buildbot-master/files/ptxdist.py
+++ b/roles/buildbot-master/files/ptxdist.py
@@ -1,5 +1,14 @@
 import os
+import string
 from buildbot.plugins import *
+from datetime import datetime
+from buildbot.config import BuilderConfig
+
+collections = {
+    "armeb-xscale": "armeb-base",
+    "i686": "i686-base",
+    "am335x": "am335x-base",
+}
 
 # Workers
 # The 'workers' list defines the set of recognized buildworkers. Each element is
@@ -12,8 +21,8 @@ workers = [
     worker.Worker("orion-am335x-slave", "pass"),
 ]
 
-worker_ptxdist_repourl = 'git@git.novatech-llc.com:andrew.cooper/workspace-ptxdist2.git'
-worker_ptxdist_branch = 'docker-build'
+worker_ptxdist_repourl = 'git@git.novatech-llc.com:Orion-ptxdist/workspace-ptxdist2'
+worker_ptxdist_branch = 'master'
 acceptance_test_repourl = 'git@git.novatech-llc.com:NovaTech-Testing/AcceptanceTests.git'
 
 # CHANGESOURCES
@@ -45,31 +54,58 @@ schedulers = [
     #                            treeStableTimer=9*60,
     #                            builderNames=["linux-3.8_am335x"]))
     schedulers.ForceScheduler(
-        name="force-ptxdist",
+        name="Force_PTXdist",
+        label="Force PTXdist Build",
         builderNames=[
-            "current_armeb_xscale",
-            "current_i686",
-            "current_am335x",
+            "force_armeb_xscale",
+            "force_i686",
+            "force_am335x",
+        ],
+        codebases=[
+            util.CodebaseParameter(
+                codebase="orion-ptxdist-workspace",
+                label="Main repository",
+                # will generate a combo box
+                branch=util.StringParameter(
+                    name="branch",
+                    default=worker_ptxdist_branch),
+                repository=util.StringParameter(
+                    name="repository",
+                    default=worker_ptxdist_repourl),
+
+                # will generate nothing in the form, but revision, repository,
+                # and project are needed by buildbot scheduling system so we
+                # need to pass a value ("")
+                revision=util.FixedParameter(name="revision", default=""),
+                project=util.FixedParameter(name="project", default="orion-ptxdist"),
+            )
         ],
         properties=[
+            util.StringParameter(
+                name="version",
+                label="distribution version",
+                default='',
+                required=True
+            ),
+            util.BooleanParameter(
+                name='release',
+                label="Make a release build",
+                default=False
+            ),
+            util.StringParameter(
+                name="packages",
+                label="space-delimited list of packages to build",
+                default='',
+                required=False,
+            ),
             util.BooleanParameter(
                 name="clobber",
                 label="Clobber build directory",
-                default=False),
+                default=False
+            ),
         ],
     ),
 
-    schedulers.Nightly(
-        name="ptxdist-nightly",
-        branch=None,
-        builderNames=[
-            "current_armeb_xscale",
-            "current_i686",
-            "current_am335x",
-        ],
-        hour=22
-    ),
-
     schedulers.Triggerable(
         name="upgrade_i686",
         builderNames=["upgrade_i686"]),
@@ -123,20 +159,74 @@ class PTXDistBuildCounter(util.LogLineObserver):
                 self.step.setProgress('packages', self.numPackages)
 
 
-class PTXDistBuild(steps.ShellCommand):
+class PTXDistBuild(steps.ShellSequence):
 
     def __init__(self, **kwargs):
         kwargs.setdefault('haltOnFailure', True)
         kwargs.setdefault('flunkOnFailure', True)
 
-        steps.ShellCommand.__init__(self, **kwargs)   # always upcall!
+        steps.ShellSequence.__init__(self, **kwargs)
+
         counter = PTXDistBuildCounter()
         self.addLogObserver('stdio', counter)
         self.progressMetrics += ('targets', 'packages')
 
-git_lock = util.MasterLock("git")
+        self.name = "PTXDist Build"
+        self.description = "building"
+        self.descriptionDone = "built"
+        self.commands = [
+            # set ptxdist build platform
+            util.ShellArg(
+                command=[
+                    "ptxdist",
+                    "platform",
+                    util.Property("platform")
+                ]),
+
+            # set ptxdist target
+            util.ShellArg(
+                command=[
+                    "ptxdist",
+                    "select",
+                    util.Property('project')
+                ]),
+
+            # run ptxdist build with build.py
+            util.ShellArg(
+                logfile="ptxdist build",
+                command=util.FlattenList([
+                    "python",
+                    "scripts/build.py",
+                    "--noclean",
+                    "--noconfirm",
+                    util.Property("version"),
+                    util.Interpolate("%(prop:release:#?|release|beta)s"),
+                    util.Transform(
+                        string.split,
+                        util.Property(
+                            "packages",
+                            default='')
+                    )
+                ])
+            ),
+
+            # set ptxdist collection
+            util.ShellArg(
+                command=[
+                    "ptxdist",
+                    "collection",
+                    util.Property("collection")
+                ]),
+
+            # If building a release, create and copy images
+            util.ShellArg(
+                command=["ptxdist", "images"]
+            ),
+
+        ]
 
-from datetime import datetime
+
+git_lock = util.MasterLock("git")
 
 
 @util.renderer
@@ -148,43 +238,70 @@ def CurrentTime(props):
     return dts
 
 
-class PTXDistFactory(util.BuildFactory):
-
-    def __init__(self, repourl, branch, platform):
-        util.BuildFactory.__init__(self)
-        self.addStep(steps.SetProperty('platform', platform))
-        self.addStep(steps.SetProperty(
-            'select', util.Interpolate("OrionLX-%(prop:platform)s-glibc")))
-        self.addStep(steps.SetProperty('timestamp', CurrentTime))
-        self.addStep(steps.SetProperty('dest', util.Interpolate(
-            "/cache/images/%(prop:buildername)s/%(prop:timestamp)s")))
-        self.addStep(steps.Git(repourl=repourl, branch=branch, mode="full",
-                               method="clobber", locks=[git_lock.access('exclusive')], retry=(360, 5)))
-        self.addStep(steps.ShellCommand(
-            command=["mkdir", "-p", util.Property('dest')]))
-        self.addStep(PTXDistBuild(
-            command=["ptxdist", "platform", util.Property("platform")]))
-        self.addStep(PTXDistBuild(
-            command=["ptxdist", "select", util.Property("select")]))
-        self.addStep(PTXDistBuild(command=["ptxdist", "go"]))
-        self.addStep(PTXDistBuild(command=["ptxdist", "make", "ipkg-push"]))
-        self.addStep(PTXDistBuild(command=["./scripts/ipkg-header"]))
-
-# current_armeb_xscale #
-current_armeb_xscale_factory = PTXDistFactory(
-    worker_ptxdist_repourl, worker_ptxdist_branch, 'armeb-xscale')
-# check out the source
-current_armeb_xscale_factory.addStep(PTXDistBuild(
-    command=["ptxdist", "collection", "armeb-base"]))
-current_armeb_xscale_factory.addStep(
-    PTXDistBuild(command=["ptxdist", "images"]))
-current_armeb_xscale_factory.addStep(steps.ShellCommand(
-    command=["cp", "platform-armeb-xscale/images/root.jffs2_64", util.Property('dest')]))
-current_armeb_xscale_factory.addStep(steps.ShellCommand(
-    command=["cp", "platform-armeb-xscale/images/root.jffs2_128", util.Property('dest')]))
-# current_armeb_xscale_factory.addStep(steps.ShellCommand(command=["./scripts/build-upgrade-test.sh"]))
-# current_armeb_xscale_factory.addStep(steps.ShellCommand(command=["curl", "--progress-bar", "-o", "/dev/null", "http://george:1234@172.16.190.70/outlet?1=CCL"]))
-# current_armeb_xscale_factory.addStep(steps.ShellCommand(command=[
+@util.renderer
+def ComputeBuildProperties(props):
+    newprops = {}
+
+    newprops['timestamp'] = ts = CurrentTime
+
+    newprops['project'] = proj = util.Interpolate(
+        "OrionLX-%(prop:platform)s-glibc"
+    )
+
+    newprops['dest'] = dest = util.Interpolate(
+        "/cache/images/%(kw:p)s", p=proj
+    )
+
+    newprops['archive'] = archive = util.Interpolate(
+        "%(kw:d)s/%(prop:machine)s.%(kw:t)s.tar.gz", d=dest, t=ts
+    )
+
+    newprops['ipkg'] = ipkg = util.Interpolate(
+        "/cache/ipkg-repository/%(kw:p)s", p=proj
+    )
+
+    newprops['collection'] = clct = collections.get(props.getProperty('platform'))
+
+    return newprops
+
+
+def isReleaseBuild(step):
+    if step.getProperty("release") is True:
+        return True
+    return False
+
+# Create build factory for ptxdist
+ptxdist_factory = util.BuildFactory()
+ptxdist_factory.addSteps([
+
+    steps.SetProperties(ComputeBuildProperties),
+
+    steps.MakeDirectory(dir=util.Property('dest')),
+
+    steps.MakeDirectory(dir=util.Property('ipkg')),
+
+    # check out the source
+    steps.Git(
+        codebase=util.Property('codebase'),
+        repourl=util.Property('repository'),
+        branch=util.Property('branch'),
+        mode=util.Interpolate("%(prop:clobber:#?|full|incremental)s"),
+        method="clobber",
+        locks=[git_lock.access('exclusive')],
+        retry=(360, 5)
+    ),
+
+    PTXDistBuild(),
+
+    steps.CopyDirectory(
+        src=util.Interpolate("platform-%(prop:platform)s/images"),
+        dest=util.Property('dest'),
+        doStepIf=isReleaseBuild
+    ),
+])
+# ptxdist_factory.addStep(steps.ShellCommand(command=["./scripts/build-upgrade-test.sh"]))
+# ptxdist_factory.addStep(steps.ShellCommand(command=["curl", "--progress-bar", "-o", "/dev/null", "http://george:1234@172.16.190.70/outlet?1=CCL"]))
+# ptxdist_factory.addStep(steps.ShellCommand(command=[
 # 	"./scripts/upgradetest.py",
 # 	"load_7_upgrade_to_8",
 # 	"/srv/tftp/root.jffs2_64",
@@ -196,8 +313,8 @@ current_armeb_xscale_factory.addStep(steps.ShellCommand(
 # 	"orionpythontests",
 # 	"orionprotocoltests",
 # 	"--orion-config", "./local-pkg/buildslave_config_armeb-xscale.tar.gz"]))
-# current_armeb_xscale_factory.addStep(trigger.Trigger(schedulerNames=['local_tests_armeb_xscale']))
-# current_armeb_xscale_factory.addStep(trigger.Trigger(schedulerNames=['remote_tests_armeb_xscale']))
+# ptxdist_factory.addStep(trigger.Trigger(schedulerNames=['local_tests_armeb_xscale']))
+# ptxdist_factory.addStep(trigger.Trigger(schedulerNames=['remote_tests_armeb_xscale']))
 
 # local_tests_armeb_xscale #
 local_tests_armeb_xscale_factory = util.BuildFactory()
@@ -217,28 +334,28 @@ remote_tests_armeb_xscale_factory.addStep(steps.ShellCommand(
     command=["py.test", "-s", "--orion=172.16.64.150", "--hub-address=172.16.64.25:4444", "--browser=chrome"], workdir='build/WebUI'))
 
 # current_i686 #
-current_i686_factory = PTXDistFactory(
-    worker_ptxdist_repourl, worker_ptxdist_branch, 'i686')
-# check out the source
-current_i686_factory.addStep(PTXDistBuild(
-    command=["ptxdist", "collection", "i686-base"]))
-current_i686_factory.addStep(PTXDistBuild(command=["ptxdist", "images"]))
-current_i686_factory.addStep(steps.ShellCommand(
-    command=["gzip", "-f", "platform-i686/images/hd.img"]))
-current_i686_factory.addStep(steps.ShellCommand(
-    command=["cp", "platform-i686/images/hd.img.gz", util.Property('dest')]))
+# current_i686_factory = PTXDistFactory(
+#     worker_ptxdist_repourl, worker_ptxdist_branch, 'i686')
+# # check out the source
+# current_i686_factory.addStep(PTXDistBuild(
+#     command=["ptxdist", "collection", "i686-base"]))
+# current_i686_factory.addStep(PTXDistBuild(command=["ptxdist", "images"]))
+# current_i686_factory.addStep(steps.ShellCommand(
+#     command=["gzip", "-f", "platform-i686/images/hd.img"]))
+# current_i686_factory.addStep(steps.ShellCommand(
+#     command=["cp", "platform-i686/images/hd.img.gz", util.Property('dest')]))
 # current_i686_factory.addStep(trigger.Trigger(schedulerNames=['upgrade_i686']))
 # current_i686_factory.addStep(steps.ShellCommand(command=["sleep", "120"]))
 # current_i686_factory.addStep(trigger.Trigger(schedulerNames=['local_tests_i686']))
 # current_i686_factory.addStep(trigger.Trigger(schedulerNames=['remote_tests_i686']))
 
 # current_am335x #
-current_am335x_factory = PTXDistFactory(
-    worker_ptxdist_repourl, worker_ptxdist_branch, 'am335x')
-# check out the source
-current_am335x_factory.addStep(PTXDistBuild(
-    command=["ptxdist", "collection", "am335x-base"]))
-current_am335x_factory.addStep(PTXDistBuild(command=["ptxdist", "images"]))
+# current_am335x_factory = PTXDistFactory(
+#     worker_ptxdist_repourl, worker_ptxdist_branch, 'am335x')
+# # check out the source
+# current_am335x_factory.addStep(PTXDistBuild(
+#     command=["ptxdist", "collection", "am335x-base"]))
+# current_am335x_factory.addStep(PTXDistBuild(command=["ptxdist", "images"]))
 # current_am335x_factory.addStep(trigger.Trigger(schedulerNames=['upgrade_am335x']))
 # current_am335x_factory.addStep(steps.ShellCommand(command=["sleep", "120"]))
 # current_am335x_factory.addStep(trigger.Trigger(schedulerNames=['local_tests_am335x']))
@@ -290,50 +407,60 @@ remote_tests_am335x_factory.addStep(steps.Git(repourl=acceptance_test_repourl, a
 remote_tests_am335x_factory.addStep(steps.ShellCommand(command=[
                                     "py.test", "-s", "--orion=172.16.190.72", "--hub-address=172.16.64.25:4444", "--browser=chrome"], workdir='build/WebUI'))
 
-from buildbot.config import BuilderConfig
-
 builders = []
 builders.append(
-    BuilderConfig(name="current_armeb_xscale",
+    BuilderConfig(name="force_armeb_xscale",
+                  workernames=["worker-ptxdist"],
+                  factory=ptxdist_factory,
+                  properties={
+                      'platform': 'armeb-xscale',
+                  }))
+builders.append(
+    BuilderConfig(name="force_i686",
                   workernames=["worker-ptxdist"],
-                  factory=current_armeb_xscale_factory))
+                  factory=ptxdist_factory,
+                  properties={
+                      'platform': 'i686',
+                  }))
+builders.append(
+    BuilderConfig(name="force_am335x",
+                  workernames=["worker-ptxdist"],
+                  factory=ptxdist_factory,
+                  properties={
+                      'platform': 'am335x',
+                  }))
+
 builders.append(
     BuilderConfig(name="local_tests_armeb_xscale",
                   workernames=["orion-armeb-xscale-slave"],
                   factory=local_tests_armeb_xscale_factory))
+builders.append(
+    BuilderConfig(name="local_tests_i686",
+                  workernames=["orion-i686-slave"],
+                  factory=local_tests_i686_factory))
+builders.append(
+    BuilderConfig(name="local_tests_am335x",
+                  workernames=["orion-am335x-slave"],
+                  factory=local_tests_am335x_factory))
+
 builders.append(
     BuilderConfig(name="remote_tests_armeb_xscale",
                   workernames=["worker-ptxdist"],
                   factory=remote_tests_armeb_xscale_factory))
 builders.append(
-    BuilderConfig(name="current_i686",
+    BuilderConfig(name="remote_tests_i686",
                   workernames=["worker-ptxdist"],
-                  factory=current_i686_factory))
+                  factory=remote_tests_i686_factory))
 builders.append(
-    BuilderConfig(name="current_am335x",
+    BuilderConfig(name="remote_tests_am335x",
                   workernames=["worker-ptxdist"],
-                  factory=current_am335x_factory))
+                  factory=remote_tests_am335x_factory))
+
 builders.append(
     BuilderConfig(name="upgrade_i686",
                   workernames=["orion-i686-slave"],
                   factory=upgrade_i686_factory))
-builders.append(
-    BuilderConfig(name="local_tests_i686",
-                  workernames=["orion-i686-slave"],
-                  factory=local_tests_i686_factory))
-builders.append(
-    BuilderConfig(name="remote_tests_i686",
-                  workernames=["worker-ptxdist"],
-                  factory=remote_tests_i686_factory))
 builders.append(
     BuilderConfig(name="upgrade_am335x",
                   workernames=["orion-am335x-slave"],
                   factory=upgrade_am335x_factory))
-builders.append(
-    BuilderConfig(name="local_tests_am335x",
-                  workernames=["orion-am335x-slave"],
-                  factory=local_tests_am335x_factory))
-builders.append(
-    BuilderConfig(name="remote_tests_am335x",
-                  workernames=["worker-ptxdist"],
-                  factory=remote_tests_am335x_factory))
diff --git a/roles/buildbot-worker-ptxdist/defaults/main.yml b/roles/buildbot-worker-ptxdist/defaults/main.yml
index 11c44750..32665ca7 100644
--- a/roles/buildbot-worker-ptxdist/defaults/main.yml
+++ b/roles/buildbot-worker-ptxdist/defaults/main.yml
@@ -64,8 +64,6 @@ hsm_host: '{{ansible_facts.default_ipv4.address}}'
 image_args:
   BUILDBOT_UID: '{{buildbot_uid}}'
   BUILDBOT_VERSION: 'v{{buildbot_version}}'
-  PTK_URI: '{{asset_host}}/610-009981-015_SW_PTK_5.3_Client_RevA.tar'
-  PTK_HSM_HOST: '{{hsm_host}}'
 
 # directory of the image source
 image_dir: '{{ docker_projects_dir }}/{{ container_hostname }}'
diff --git a/roles/buildbot-worker-ptxdist/tasks/main.yml b/roles/buildbot-worker-ptxdist/tasks/main.yml
index b85341a0..9b0bce1f 100644
--- a/roles/buildbot-worker-ptxdist/tasks/main.yml
+++ b/roles/buildbot-worker-ptxdist/tasks/main.yml
@@ -55,6 +55,11 @@
     repo: '{{ image_repo }}'
     version: master
     dest: '{{ image_dir }}'
+    force: yes
+
+- template:
+    src: pydistutils.cfg.j2
+    dest: '{{ image_dir }}/buildbot/.pydistutils.cfg'
 
 - name: Create buildbot worker image
   docker_image:
